### PR TITLE
Added new changes in BinarySearchTree

### PR DESCRIPTION
**Pull request by mohanprasathsj**
Riyad,
I have added the following things
1. ValidateTree - Gives the whole tree is valid or not (check whether
child's parent has that as child in either Left or Right.)
2. PreOrderTraversal - Helps to get the all the nodes in PreOrder. This
helps to get the overall Tree Structure
3. Corrected the Re-Orient functionaly in Transpose
4. Added Test for the above changes

Thanks for your time. Let me know if you got any queries regarding the above changes. 

Regards, 
Mohan Prasath S J

### CODE REVIEW

File: data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs [hunk 1]
```diff
@@ -415,6 +415,37 @@ private bool IsInRange(TKey start, TKey end, Node<TKey, TValue> node)
...
+        /// <returns></returns>
+        private List<KeyValuePair<TKey, TValue>> GetAllNodesInPreOrder(Node<TKey, TValue> root)
+        {
+            Contract.Requires<ArgumentNullException>(root != null);
+            Contract.Requires<ArgumentNullException>(this.root != null);
+            Contract.Ensures(Contract.Result<List<KeyValuePair<TKey, TValue>>>() != null);
+
+            var list = new List<KeyValuePair<TKey, TValue>>();
+
+            if (root.Equals(new NullNode<TKey, TValue>()))              //If root node is a NullNode
```
Comment (riyadparvez) [inline data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs]: It creates a `NullNode` object every time, better create a static `NullNode` object member for the class, that way you don't pay the penalty of creating new object everytime.

File: data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs [hunk 2]
```diff
@@ -415,6 +415,37 @@ private bool IsInRange(TKey start, TKey end, Node<TKey, TValue> node)
         }
 
         /// <summary>
+        /// Traverse through all the nodes from root node and write the data in Pre-Order Traversal. This returns the result in ascending order
+        /// </summary>
+        /// <param name="root"></param>
+        /// <returns></returns>
+        private List<KeyValuePair<TKey, TValue>> GetAllNodesInPreOrder(Node<TKey, TValue> root)
```
Comment (riyadparvez) [inline data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs]: Make the return value `IList`.

File: data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs (hunk 1 above)
Comment (mohanprasathsj) [inline data-structures-csharp/data-structures-csharp/BinarySearchTree/BinarySearchTree.cs]: That's a good idea. Will work on that.

**Comment by mohanprasathsj** [comment]
Converted NullNode to Singleton and changed List to IList. You can merge to your code. Let me know if there is any change. 

Thanks. :)

**Comment by riyadparvez** [comment]
Thanks for the PR!
Can you add a travis build for this project? I don't use Windows anymore, so it's hard for me to see if the code is compiling.

**Comment by mohanprasathsj** [comment]
HI Riyad, 
I haven't tried travis build before. I will check on that this weekend.
Thanks:)

**Pull request merged**